src/util/locking/RedisLocker.ts

```diff
@@ -0,0 +1,225 @@
+  retryJitter?: number;
+}
+
+const attemptDefaults: Required<AttemptSettings> = { retryCount: -1, retryDelay: 50, retryJitter: 30 };
+
+// Internal prefix for Redis keys;
+const PREFIX_RW = '__RW__';
+const PREFIX_LOCK = '__L__';
+
+/**
+ * A Redis Locker that can be used as both:
+ *  *  a Read Write Locker that uses a (single) Redis server to store the locks and counts.
+ *  *  a Resource Locker that uses a (single) Redis server to store the lock.
+ * This solution should be process-safe. The only references to locks are string keys
+ * derived from identifier paths.
+ *
+ * The Read Write algorithm roughly goes as follows:
+ *  * Acquire a read lock: allowed as long as there is no write lock. On acquiring the read counter goes up.
+ *  * Acquire a write lock: allowed as long as there is no other write lock AND the read counter is 0.
+ *  * Release a read lock: decreases the read counter with 1
+ *  * Release a write lock: unlocks the write lock
+ *
+ * The Resource locking algorithm uses a single mutex/lock.
+ *
+ * The operations like checking for a write lock AND read count are considered atomic because of being written
+ * in one Lua script. Such a lua scripts is uploaded to Redis as a single new command.
+ *
+ * Redis executes its operations single threaded, as such, each operation on its own can be considered atomic
+ * (one lua script is considered one operation.)
+ */
+export class RedisLocker implements ReadWriteLocker, ResourceLocker, Finalizable {
+  protected readonly logger = getLoggerFor(this);
+
+  private readonly redis: Redis;
+  private readonly redisRw: RedisReadWriteLock;
+  private readonly redisLock: RedisResourceLock;
+  private readonly attemptSettings: Required<AttemptSettings>;
+
+  public constructor(redisClient = '127.0.0.1:6379', attemptSettings: AttemptSettings = {}) {
+    this.redis = this.createRedisClient(redisClient);
+    this.attemptSettings = { ...attemptDefaults, ...attemptSettings };
+
+    // Register lua scripts
+    this.redis.defineCommand('acquireReadLock', {
+      numberOfKeys: 1,
+      lua: SCRIPTS.acquireReadLock,
+    });
+    this.redis.defineCommand('acquireWriteLock', {
+      numberOfKeys: 1,
+      lua: SCRIPTS.acquireWriteLock,
+    });
+    this.redis.defineCommand('releaseReadLock', {
+      numberOfKeys: 1,
+      lua: SCRIPTS.releaseReadLock,
+    });
+    this.redis.defineCommand('releaseWriteLock', {
+      numberOfKeys: 1,
+      lua: SCRIPTS.releaseWriteLock,
+    });
+    this.redis.defineCommand('acquireLock', {
+      numberOfKeys: 1,
+      lua: SCRIPTS.acquireLock,
+    });
+    this.redis.defineCommand('releaseLock', {
+      numberOfKeys: 1,
+      lua: SCRIPTS.releaseLock,
+    });
+
+    // @ts-expect-error Use custom interface type including the ReadWrite Lock operations just added via lua.
+    this.redisRw = this.redis;
+    // @ts-expect-error Use custom interface type including the Resource Lock operations just added via lua.
+    this.redisLock = this.redis;
```

Comment:
We should never ignore typescript errors. Should be `this.redisRw = this.redis as RedisReadWriteLock`, or `this.redisRw = this.redis as unknown as RedisReadWriteLock` should the first not be accepted.

Reply:
Will try, the any was after some struggling with typing